q2_sample_classifier/classify.py

```diff
@@ -371,6 +371,12 @@ def confusion_matrix(output_dir: str,
     if vmax == 'auto':
         vmax = None
 
+    try:
+        dtype = type(truth.to_series()[0])
+        predictions = predictions.astype(dtype)
+    except ValueError:
```

Comment:
If we hit this exception, we are guaranteed to hit the exception mentioned in #179 later down the road aren't we? Because it means the types were different and cannot be coerced to be the same doesn't it? So do we just want to raise that exception here? I didn't do it initially because I'm not 100% confident that is the case.

Reply:
yes — though I am not sure we could ever really hit this exception, since `truth` should always be `object`, since it is a `CategoricalMetadataColumn` and QIIME 2 will already be validating that... right? 

so maybe we do not even need to check the dtype necessarily, just cast `prediction` as an `object`?

but if you stick with checking the `dtype` as you have done here, I'd suggest:
1. `truth.to_series().dtype` seems to make more sense than `type(truth.to_series()[0])`
2. `except` should raise an error if you can figure out a way to hit the exception! otherwise take this out of a try/except.

Reply:
Casting `prediction` to `object` without the `dtype` checking and all that doesn't seem to work. `truth.to_series().dtype` doesn't get the actual type it just gets `object` which, as stated above, doesn't really seem to work. And yeah I'm not really seeing a way for that `except` to actually get hit. The idea was if we couldn't coerce the `dtypes` we would hit the error from #179. But yeah I'm not particularly seeing any way to hit that exception.

Reply:
The except gets hit by the `test_confusion_matrix_class_overlap_error` test where . So it gets hit in the scenario where the `predictions` are strings and the truth values are ints. This causes those two lines of code to attempt to coerce the string "a" into a numerical value which fails. The except just throws those changes away causing the later exception (saying there is no overlap) to be hit. The question is, do we want to warn them that something is wrong in that except block? Or just let them hit the error that is currently being raised?

Reply:
`test_confusion_matrix_class_overlap_error` is not _intended_ to hit that exception, and is a relic of an earlier day before bug #179 was discovered.

the goal of `test_confusion_matrix_class_overlap_error` is to hit a different warning, when truth and prediction values do not overlap. Unless if we cast truth as str (see below), this test should be changed so that the truth values are strings.

So with those changes it sounds like my point above still stands: we should not be hitting the exception on line 377, and both truth and predictions should always be objects/str since they are `CategoricalMetadataColumn`. Maybe we should just explicitly cast truth and predictions both as str instead of checking the dtype.